src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs

```diff
@@ -0,0 +1,124 @@
+    public class UdfComplexTypesTests
+    {
+        private readonly SparkSession _spark;
+        private readonly DataFrame _df;
+
+        public UdfComplexTypesTests(SparkFixture fixture)
+        {
+            _spark = fixture.Spark;
+            _df = _spark
+                .Read()
+                .Json($"{TestEnvironment.ResourceDirectory}people_types.json");
+        }
+
+        // UDF that takes in Array type.
+        [Fact]
+        public void TestUdfWithArrayType()
+        {
+            //ArrayList works for this type.
+            Func<Column, Column> udfInt = Udf<int[], string>(
+                    (arr) =>
+                    {
+                        return string.Join(',', arr);
+                    });
+
+            Assert.Throws<Exception>(() => _df.Select(udfInt(_df["ages"])).Collect().ToArray());
+
+            Func<Column, Column> udfString = Udf<string[], string>(
+                    (arr) =>
+                    {
+                        return string.Join(',', arr);
+                    });
+
+            Assert.Throws<Exception>(() => _df.Select(udfString(_df["cars"])).Collect().ToArray());
+        }
+
+        // UDF that returns Array type.
+        [Fact]
+        public void TestUdfWithReturnTypeAsArray()
+        {
+            Func<Column, Column> udf = Udf<string, string[]>(
+                    (str) =>
+                    {
+                        return new string[] { str, str + str };
+                    });
+
+            Assert.Throws<NotImplementedException>(() => _df.Select(udf(_df["name"])).Collect().ToArray());
+        }
+
+        // UDF that takes in Map type.
+        [Fact]
+        public void TestUdfWithMapType()
+        {
+            Func<Column, Column> udf = Udf<IDictionary<string, string>, string>(
+                    (dict) =>
+                    {
+                        return dict.Count.ToString();
+                    });
+
+            Assert.Throws<Exception>(() => _df.Select(udf(_df["info"])).Collect().ToArray());
+        }
+
+        // UDF that returns Map type.
+        [Fact]
+        public void TestUdfWithReturnTypeAsMap()
+        {
+            Func<Column, Column> udf = Udf<string, IDictionary<string, string>>(
+                (str) =>
+                {
+                    return new Dictionary<string, string> { { str, str } };
+                });
+
+            Assert.Throws<NotImplementedException>(() => _df.Select(udf(_df["name"])).Collect().ToArray());
+        }
+
+        // UDF that takes in Row type.
+        [Fact]
+        public void TestUdfWithRowType()
+        {
+            Func<Column, Column> udf = Udf<Row, string>(
+                (row) =>
+                {
+                    string city = row.GetAs<string>("city");
+                    string state = row.GetAs<string>("state");
+                    return $"{city},{state}";
+                });
+
+            Row[] rows = _df.Select(udf(_df["info"])).Collect().ToArray();

```

Comment:
Why does this need to be an array?  Couldn't you stick with `Collect()` and iterate through the `IEnumerable` ?